atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java

```diff
@@ -0,0 +1,214 @@
+    }
+
+    @VisibleForTesting
+    SequenceAndBound forceAgreedState(long seq, @Nullable Long oldState) {
+        if (seq <= PaxosAcceptor.NO_LOG_ENTRY) {
+            return ImmutableSequenceAndBound.of(PaxosAcceptor.NO_LOG_ENTRY, 0L);
+        }
+
+        Optional<SequenceAndBound> state = getLearnedState(seq);
+        if (state.isPresent()) {
+            return state.get();
+        }
+
+        while (true) {
+            try {
+                byte[] value = proposer.propose(seq, oldState == null ? null : PtBytes.toBytes(oldState));
+                // propose must never be null.  We only pass in null for things we know are agreed upon already.
+                Preconditions.checkNotNull(value, "Proposed value can't be null, but was in sequence %s", seq);
+                return ImmutableSequenceAndBound.of(seq, PtBytes.toLong(value));
+            } catch (PaxosRoundFailureException e) {
+                log.info("failed during propose", e);
+                long backoffTime = getRandomBackoffTime();
+                try {
+                    Thread.sleep(backoffTime);
+                } catch (InterruptedException e1) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+    }
+
+    private Optional<SequenceAndBound> getLearnedState(long seq) {
+        if (seq <= PaxosAcceptor.NO_LOG_ENTRY) {
+            return Optional.of(ImmutableSequenceAndBound.of(PaxosAcceptor.NO_LOG_ENTRY, 0L));
+        }
+        List<PaxosLong> responses = PaxosQuorumChecker.collectQuorumResponses(
+                ImmutableList.copyOf(learners),
+                l -> getLearnedValue(seq, l),
+                1,
+                executor,
+                PaxosQuorumChecker.DEFAULT_REMOTE_REQUESTS_TIMEOUT_IN_SECONDS,
+                true);
+        if (responses.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(ImmutableSequenceAndBound.of(seq, responses.iterator().next().getValue()));
+    }
+
+    private static PaxosLong getLearnedValue(long seq, PaxosLearner learner) {
+        PaxosValue value = learner.getLearnedValue(seq);
+        if (value == null) {
+            throw new NoSuchElementException();
+        }
+        return ImmutablePaxosLong.of(PtBytes.toLong(value.getData()));
+    }
+
+    @Override
+    public synchronized void storeUpperLimit(long limit) throws MultipleRunningTimestampServiceError {
+        Preconditions.checkArgument(agreedState == null || limit >= agreedState.getBound());
+        long newSeq = agreedState == null ? 0 : agreedState.getSeqId() + 1;
+        while (true) {
+            try {
+                proposer.propose(newSeq, PtBytes.toBytes(limit));
+                PaxosValue value = knowledge.getLearnedValue(newSeq);
+                if (!value.getLeaderUUID().equals(proposer.getUUID())) {
+                    throw new MultipleRunningTimestampServiceError("concurrent proposal: " + value);
```

Comment:
Okay. Actually the PaxosValue has quite a lot of information, the only additional info that I think we have here is our UUID and the value we tried to propose. I've added that, cleaned this up a fair bit, in any case, and made it more readable.

```java
String errorMsg = String.format(
        "Timestamp limit changed from under us for sequence '%s' (leader with UUID '%s' changed"
        + " it, our UUID is '%s'). This suggests that another timestamp store for this"
        + " namespace is running. The offending bound was '%s'; we tried to propose"
        + " a bound of '%s'. (The offending Paxos value was '%s'.)",
        newSeq,
        value.getLeaderUUID(),
        proposer.getUUID(),
        PtBytes.toLong(value.getData()),
        limit,
        value);
```